source/common/router/config_impl.cc

```diff
@@ -1522,25 +1537,42 @@ RouteConstSharedPtr ConfigImpl::route(const RouteCallback& cb,
   return route_matcher_->route(cb, headers, stream_info, random_value);
 }
 
-namespace {
+RouteSpecificFilterConfigConstSharedPtr PerFilterConfigs::createRouteSpecificFilterConfig(
+    const std::string& name, const ProtobufWkt::Any& typed_config,
+    const ProtobufWkt::Struct& config, const OptionalHttpFilters& optional_http_filters,
+    Server::Configuration::ServerFactoryContext& factory_context,
+    ProtobufMessage::ValidationVisitor& validator) {
+  auto factory = Envoy::Config::Utility::getAndCheckFactoryByName<
+      Server::Configuration::NamedHttpFilterConfigFactory>(name, optional_http_filters.find(name) !=
+                                                                     optional_http_filters.end());
+  if (factory == nullptr) {
+    ENVOY_LOG(warn, "Can't find a registered implementation for http filter '{}'", name);
+    return nullptr;
+  }
 
-RouteSpecificFilterConfigConstSharedPtr
-createRouteSpecificFilterConfig(const std::string& name, const ProtobufWkt::Any& typed_config,
-                                const ProtobufWkt::Struct& config,
-                                Server::Configuration::ServerFactoryContext& factory_context,
-                                ProtobufMessage::ValidationVisitor& validator) {
-  auto& factory = Envoy::Config::Utility::getAndCheckFactoryByName<
-      Server::Configuration::NamedHttpFilterConfigFactory>(name);
-  ProtobufTypes::MessagePtr proto_config = factory.createEmptyRouteConfigProto();
+  ProtobufTypes::MessagePtr proto_config = factory->createEmptyRouteConfigProto();
   Envoy::Config::Utility::translateOpaqueConfig(typed_config, config, validator, *proto_config);
-  return factory.createRouteSpecificFilterConfig(*proto_config, factory_context, validator);
+  auto object = factory->createRouteSpecificFilterConfig(*proto_config, factory_context, validator);
+  if (object == nullptr) {
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.check_unsupported_typed_per_filter_config")) {
+      throw EnvoyException(
+          fmt::format("The filter {} doesn't support virtual host-specific configurations", name));
```

Comment:
Should we ignore the unsupported typed per filter config for the optional http filter also? I assume there will be the same upgrade issue for the new filter. like, the filter added typed per filter config in the new version, the old version can't support it. Although there is a runtime flag here, the is_optional flag feels better.

Reply:
Yes, I think that if `is_optional` is set, we should ignore unknown per-route filter configs.

Reply:
got it, thanks